File: LLMgeneratedCode/4o-3.py
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
from sklearn.linear_model import Ridge
from sklearn.metrics import mean_squared_error

# Load training data
feature_train = pd.read_csv("df_training.csv")
metadata_train = pd.read_csv("ano_training.csv")

# Load test data
feature_test = pd.read_csv("df_test.csv")
metadata_test = pd.read_csv("ano_test.csv")

# Transpose feature data so samples are rows and features are columns
feature_train = feature_train.set_index(feature_train.columns[0]).T
feature_test = feature_test.set_index(feature_test.columns[0]).T

# Ensure the index (Sample_ID) matches between features and metadata
metadata_train = metadata_train.set_index("Sample_ID").loc[feature_train.index]
metadata_test = metadata_test.set_index("Sample_ID").loc[feature_test.index]

# Extract gestational age from metadata
y_train = metadata_train["GA"].values
y_test = metadata_test["GA"].values

# Convert feature data to numpy arrays
X_train = feature_train.values
X_test = feature_test.values

# Train a ridge regression model (L2 regularization for high-dimensional data)
model = Ridge(alpha=1.0)
model.fit(X_train, y_train)

# Predict on test set
y_pred = model.predict(X_test)

# Compute RMSE
rmse = np.sqrt(mean_squared_error(y_test, y_pred))
print(f"Test RMSE: {rmse:.2f}")

# Scatter plot of actual vs predicted GA
plt.figure(figsize=(6, 6))
plt.scatter(y_test, y_pred, alpha=0.5, label=f"RMSE: {rmse:.2f}")
plt.plot([min(y_test), max(y_test)], [min(y_test), max(y_test)], 'r--', label="Ideal Fit")
plt.xlabel("Actual GA")
plt.ylabel("Predicted GA")
plt.legend()
plt.title("Predicted vs Actual Gestational Age")
plt.savefig("./out2/4o-3.py.pdf", format="pdf")
df = pd.DataFrame({'y_test': y_test, 'y_pred': y_pred})
df.to_csv("./out2/4o-3.py.csv", index=False)

File: LLMgeneratedCode/4o-3.R
library(glmnet)
library(readr)
library(dplyr)
library(ggplot2)
library(caret)

# Load training data
feature_train <- read_csv("df_training.csv")
metadata_train <- read_csv("ano_training.csv")

# Load test data
feature_test <- read_csv("df_test.csv")
metadata_test <- read_csv("ano_test.csv")

# Transpose feature data so that samples are rows
feature_train <- as.data.frame(t(feature_train))
colnames(feature_train) <- feature_train[1, ]
feature_train <- feature_train[-1, ]
feature_train <- feature_train %>% mutate(Sample_ID = rownames(feature_train))

feature_test <- as.data.frame(t(feature_test))
colnames(feature_test) <- feature_test[1, ]
feature_test <- feature_test[-1, ]
feature_test <- feature_test %>% mutate(Sample_ID = rownames(feature_test))

# Convert feature data to numeric
feature_train[,-ncol(feature_train)] <- lapply(feature_train[,-ncol(feature_train)], as.numeric)
feature_test[,-ncol(feature_test)] <- lapply(feature_test[,-ncol(feature_test)], as.numeric)

# Merge with metadata
train_data <- merge(metadata_train, feature_train, by = "Sample_ID")
test_data <- merge(metadata_test, feature_test, by = "Sample_ID")

# Extract gestational age (GA) and feature matrix
GA_train <- train_data$GA
GA_test <- test_data$GA

X_train <- as.matrix(train_data[ , !(colnames(train_data) %in% c("Sample_ID", "GA"))])
X_test <- as.matrix(test_data[ , !(colnames(test_data) %in% c("Sample_ID", "GA"))])

set.seed(123)
# Fit elastic net model
cv_model <- cv.glmnet(X_train, GA_train, alpha = 0.5)

# Predict on test set
predictions <- predict(cv_model, X_test, s = "lambda.min")

# Compute RMSE
rmse <- sqrt(mean((GA_test - predictions)^2))
print(paste("RMSE on test set:", round(rmse, 2)))
write.csv(data.frame(y_test=GA_test,y_pred=predictions),file="./out2/4o-3.R.csv")

# Plot predicted vs actual GA
plot_df <- data.frame(Actual = GA_test, Predicted = predictions)
plot <- ggplot(plot_df, aes(x = Actual, y = Predicted)) +
  geom_point(alpha = 0.5) +
  geom_smooth(method = "lm", color = "blue", se = FALSE) +
  ggtitle("Predicted vs Actual Gestational Age") +
  xlab("Actual GA") +
  ylab("Predicted GA") +
  annotate("text", x = min(GA_test), y = max(predictions), label = paste("RMSE:", round(rmse, 2)), hjust = 0, size = 5, color = "red") +
  theme_minimal()

print(plot)
